src/axom/primal/operators/detail/intersect_bezier_impl.hpp

```diff
@@ -0,0 +1,204 @@
+// Copyright (c) 2017-2019, Lawrence Livermore National Security, LLC and
+// other Axom Project Developers. See the top-level COPYRIGHT file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+
+/*!
+ * \file intersect_bezier_impl.hpp
+ *
+ * This file provides helper functions for testing the intersection
+ * of Bezier curves
+ */
+
+#ifndef PRIMAL_INTERSECT_BEZIER_IMPL_HPP_
+#define PRIMAL_INTERSECT_BEZIER_IMPL_HPP_
+
+#include "axom/primal/geometry/Point.hpp"
+#include "axom/primal/geometry/BoundingBox.hpp"
+#include "axom/primal/geometry/BezierCurve.hpp"
+
+#include "axom/primal/operators/intersect.hpp"
+#include "axom/primal/operators/detail/intersect_impl.hpp"
+
+#include <vector>
+
+namespace axom
+{
+namespace primal
+{
+namespace detail
+{
+
+//---------------------------- FUNCTION DECLARATIONS ---------------------------
+
+/*!
+ * \brief Recursive function to find the intersections between two Bezier curves
+ *
+ * \param [in] c1 The first input curve
+ * \param [in] c2 The second input curves
+ * \param [out] sp Parametric coordinates of intersections in \a c1
+ * \param [out] tp Parametric coordinates of intersections in \a c2
+ * \param [in] sq_tol The squared tolerance parameter for determining if a
+ * Bezier curve is linear
+ * \param [in] order1 The order of \a c1
+ * \param [in] order2 The order of \a c2
+ * \param s_offset The offset in parameter space for \a c1
+ * \param s_scale The scale in parameter space for \a c1
+ * \param t_offset The offset in parameter space for \a c2
+ * \param t_scale The scale in parameter space for \a c2
+ *
+ * Bezier curves can only intersect when their bounding boxes intersect.
+ * The base case of the recursion is when we can approximate the curves as
+ * line segments, where we directly find their intersections. Otherwise,
+ * check for intersections recursively after bisecting one of the curves.
+ *
+ * \note A BezierCurve is parametrized in [0,1). The scale and offset parameters
+ * are used to track the local curve parameters during subdivisions
+ *
```

Comment:
Please note what happens when the Bezier curves are identical or have sub-sections that are identical.  Also note what happens when the Bezier curves have sections that are very close, such as a tangent intersection.

Reply:
I see that you documented the user-facing functions to point out that they handle non-tangent intersections of multiplicity 1.  That is OK with me.